src/ophyd_async/core/_providers.py

```diff
@@ -17,43 +20,155 @@ class DirectoryInfo:
+    def __init__(self, uuid_call_func: callable = uuid.uuid4):
+        self._uuid_call_func = uuid_call_func
+        self._uuid_namespace = None
+        self._uuid_name = None
+
+    def specify_uuid_namespace(self, namespace, name):
+        self._uuid_namespace = namespace
+        self._uuid_name = name
+
+    def __call__(self) -> str:
+        if self._uuid_call_func in [uuid.uuid3, uuid.uuid5]:
+            if self._uuid_namespace is None or self._uuid_name is None:
+                raise ValueError(
+                    f"To use {self._uuid_call_func} to generate UUID filenames,"
+                    " UUID namespace and name must be set!"
+                )
+            uuid_str = self._uuid_call_func(self._uuid_namespace, self._uuid_name)
+        else:
+            uuid_str = self._uuid_call_func()
+        return f"{uuid_str}"
+
+
+class AutoIncrementFilenameProvider(FilenameProvider):
+    def __init__(
+        self,
+        base_filename: str = "",
+        max_digits: int = 5,
+        starting_value: int = 0,
+        increment: int = 1,
+        inc_delimeter: str = "_",
+    ):
+        self._base_filename = base_filename
+        self._max_digits = max_digits
+        self._current_value = starting_value
+        self._increment = increment
+        self._inc_delimeter = inc_delimeter
+
+    def __call__(self):
+        if len(str(self._current_value)) > self._max_digits:
+            raise ValueError(
+                f"Auto incrementing filename counter \
+                  exceeded maximum of {self._max_digits} digits!"
+            )
+
+        padded_counter = str(self._current_value).rjust(self._max_digits, "0")
+
+        filename = f"{self._base_filename}{self._inc_delimeter}{padded_counter}"
+
+        self._current_value += self._increment
+        return filename
+
+
+class StaticPathProvider(PathProvider):
     def __init__(
         self,
+        filename_provider: FilenameProvider,
         directory_path: Union[str, Path],
+        resource_dir: Path = Path("."),
         filename_prefix: str = "",
         filename_suffix: str = "",
-        resource_dir: Optional[Path] = None,
+        create_dir_depth: int = -3,
     ) -> None:
-        if resource_dir is None:
-            resource_dir = Path(".")
-        if isinstance(directory_path, str):
-            directory_path = Path(directory_path)
-        self._directory_info = DirectoryInfo(
-            root=directory_path,
-            resource_dir=resource_dir,
-            prefix=filename_prefix,
-            suffix=filename_suffix,
+        self._filename_provider = filename_provider
+        self._directory_path = directory_path
+        self._resource_dir = resource_dir
+        self._create_dir_depth = create_dir_depth
+        self._prefix = filename_prefix
+        self._suffix = filename_suffix
+
+    def __call__(self, device_name=None) -> PathInfo:
+        filename = f"{self._prefix}{self._filename_provider()}{self._suffix}"
```

Comment:
filenameprovider should be handling prefixing/suffixing, StaticFileNameProvider -> StaticFormatFileNamer? init(prefix: Optional[str], suffix: Optional[str]) ?